# Q1 - Grid-to-Finish Delta - Data Validation and Sanity Checks

In [7]:
import pandas as pd

# read csv file
df_results = pd.read_csv('/Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/grid-to-finish.csv')

# dataframe basic info
print(df_results.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 604 entries, 0 to 603
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   race_id          604 non-null    int64 
 1   gp_year          604 non-null    int64 
 2   gp_name          604 non-null    object
 3   gp_round         604 non-null    int64 
 4   driver_name      604 non-null    object
 5   constructor      604 non-null    object
 6   constructor_ref  604 non-null    object
 7   is_williams      604 non-null    bool  
 8   start_position   604 non-null    int64 
 9   final_position   604 non-null    int64 
 10  grid_delta       604 non-null    int64 
dtypes: bool(1), int64(6), object(4)
memory usage: 47.9+ KB
None


## Summary of processed dataset 'grid-to-finish.csv'

- Filepath: /Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/grid-to-finish.csv *(potentially fix from absolute to relative path later?)*
- Range: 604 entries, 0 to 603.
- Columns: 11
- Data types: int64(6), object(4), bool(1) *(objects are strings)*
- Memory usage: 47.9+ KB 

## Column data types

In [8]:
print(df_results.dtypes)

race_id             int64
gp_year             int64
gp_name            object
gp_round            int64
driver_name        object
constructor        object
constructor_ref    object
is_williams          bool
start_position      int64
final_position      int64
grid_delta          int64
dtype: object


## Missing or null values

In [9]:
df_results.isnull().sum() # No nulls present across the dataset!

race_id            0
gp_year            0
gp_name            0
gp_round           0
driver_name        0
constructor        0
constructor_ref    0
is_williams        0
start_position     0
final_position     0
grid_delta         0
dtype: int64

## Check for duplicates

In [10]:
df_results.duplicated().sum() # no duplicates found

0

## Summary statistics

In [12]:
df_results.describe()

race_id      gp_year    gp_round  start_position  final_position  \
count   604.000000   604.000000  604.000000      604.000000      604.000000   
mean    983.165563  2017.233444   10.769868       11.882450       10.736755   
std      28.197934     1.312072    5.885166        4.605256        3.981538   
min     926.000000  2015.000000    1.000000        2.000000        3.000000   
25%     960.000000  2016.000000    6.000000        8.000000        7.000000   
50%     984.000000  2017.000000   11.000000       11.000000       11.000000   
75%    1007.000000  2018.000000   16.000000       16.000000       14.000000   
max    1030.000000  2019.000000   21.000000       22.000000       22.000000   

       grid_delta  
count  604.000000  
mean     1.145695  
std      3.774133  
min    -17.000000  
25%     -1.000000  
50%      1.000000  
75%      4.000000  
max     13.000000

From this we can roughly tell that, 
- Mean grid to finish delta is a gain of 1.14 places, median a gain of 1 place.
- Standard deviation of delta is 3.77 places
- The most places gained between 2015 and 2019, between Williams and its midfield competitors, is 13 places.
- The most places lost, for the same period and teams, is 17 places. 

In [13]:
df_results.to_csv('/Users/frankdong/Documents/Analytics Local/williams-racing-strategies/processed_data/grid-to-finish_validated.csv', index=False)

## Validation conclusion
- No null values found
- Column data types are correct
- Grid delta logic is confirmed
- Proceed with feature engineering using the provided CSV data in 'grid-to-finish_validated.csv'